packages/docker-engine/0001-read-default-oci-spec-from-file.patch

```diff
@@ -0,0 +1,170 @@
++	BaseRuntimeSpec      string                   `json:"base-runtime-spec,omitempty"`
+ 	// ResolvConf is the path to the configuration of the host resolver
+ 	ResolvConf string `json:"resolv-conf,omitempty"`
+ 	Rootless   bool   `json:"rootless,omitempty"`
+diff --git a/daemon/oci_linux.go b/daemon/oci_linux.go
+index a5a5acf..53908ce 100644
+--- a/daemon/oci_linux.go
++++ b/daemon/oci_linux.go
+@@ -2,6 +2,7 @@ package daemon // import "github.com/docker/docker/daemon"
+
+ import (
+ 	"context"
++	"encoding/json"
+ 	"fmt"
+ 	"io/ioutil"
+ 	"os"
+@@ -24,6 +25,7 @@ import (
+ 	"github.com/docker/docker/pkg/stringid"
+ 	"github.com/docker/docker/rootless/specconv"
+ 	volumemounts "github.com/docker/docker/volume/mounts"
++	units "github.com/docker/go-units"
+ 	"github.com/moby/sys/mount"
+ 	"github.com/moby/sys/mountinfo"
+ 	"github.com/opencontainers/runc/libcontainer/apparmor"
+@@ -45,8 +47,13 @@ func WithRlimits(daemon *Daemon, c *container.Container) coci.SpecOpts {
+
+ 		// We want to leave the original HostConfig alone so make a copy here
+ 		hostConfig := *c.HostConfig
++
++		//Merge defaults in Base Spec file with Daemon defaults
++		daemon.mergeBaseSpecUlimits(s)
++
+ 		// Merge with the daemon defaults
+ 		daemon.mergeUlimits(&hostConfig)
++
+ 		for _, ul := range hostConfig.Ulimits {
+ 			rlimits = append(rlimits, specs.POSIXRlimit{
+ 				Type: "RLIMIT_" + strings.ToUpper(ul.Name),
+@@ -158,8 +165,12 @@ func WithApparmor(c *container.Container) coci.SpecOpts {
+ // WithCapabilities sets the container's capabilties
+ func WithCapabilities(c *container.Container) coci.SpecOpts {
+ 	return func(ctx context.Context, _ coci.Client, _ *containers.Container, s *coci.Spec) error {
++		defCaps := caps.DefaultCapabilities()
++		defCaps = append(defCaps, s.Process.Capabilities.Bounding...)
++		defCaps = getUniqueCapabilities(defCaps)
++
+ 		capabilities, err := caps.TweakCapabilities(
+-			caps.DefaultCapabilities(),
++			defCaps,
+ 			c.HostConfig.CapAdd,
+ 			c.HostConfig.CapDrop,
+ 			c.HostConfig.Privileged,
+@@ -171,6 +182,19 @@ func WithCapabilities(c *container.Container) coci.SpecOpts {
+ 	}
+ }
+
++func getUniqueCapabilities(defCaps []string) []string {
++	occurred := map[string]bool{}
++	var uniqueCaps []string
++	for c := range defCaps {
++		if occurred[defCaps[c]] != true {
++			occurred[defCaps[c]] = true
++			uniqueCaps = append(uniqueCaps, defCaps[c])
++		}
++	}
++
++	return uniqueCaps
++}
++
+ func resourcePath(c *container.Container, getPath func() (string, error)) (string, error) {
+ 	p, err := getPath()
+ 	if err != nil {
+@@ -1014,6 +1038,23 @@ func (daemon *Daemon) createSpec(c *container.Container) (retSpec *specs.Spec, e
+ 		opts []coci.SpecOpts
+ 		s    = oci.DefaultSpec()
+ 	)
++
++	if daemon.configStore.BaseRuntimeSpec != "" {
++		baseSpec, errReadfile := os.ReadFile(daemon.configStore.BaseRuntimeSpec)
++		if errReadfile != nil {
++			fmt.Print("Can't find base OCI spec %q", daemon.configStore.BaseRuntimeSpec)
```

Comment:
Need to be able to break out from here if there was an error. Otherwise this just prints the message and continues on where it tries to use `baseSpec` on line 130. Though I suppose that will just fail there, so technically this does work.

Reply:
Ohh yeah good point, or may be break from if loop and use default spec so atleast it will not break and create the containers, it could be a little hard to find later what went wrong that it does not shows the settings in Base-spec-file. So correct if we are ok for error we can leave it as is, else break if and use default.